lib/AST/ShouldPrintForParseableInterface.cpp

```diff
@@ -83,12 +92,63 @@ class FindReferencedNonPublicTypeDecls: public ASTWalker {
   FindTypeDecls findTypeDecls;
 public:
   FindReferencedNonPublicTypeDecls(TypeDecls &foundDecls):
-    findTypeDecls(foundDecls) {}
+    findTypeDecls(foundDecls, workList) {}
+
+  void search(ModuleDecl *module) {
+    SmallVector<Decl *, 8> topLevelDecls;
+    module->getTopLevelDecls(topLevelDecls);
+    for (auto decl : topLevelDecls) {
+      auto vd = dyn_cast<ValueDecl>(decl);
+      if (!vd || !isPublicOrUsableFromInline(vd))
+        continue;
+      workList.push_back(vd);
+    }
+
+    while (!workList.empty()) {
+      Decl *d = workList.pop_back_val();
+      d->walk(*this);
+    }
+  }
+
+  void visitAbstractStorageDecl(AbstractStorageDecl *decl) {
+    if (!contributesToParentTypeStorage(decl))
+      return;
+    decl->getInterfaceType().walk(findTypeDecls);
+  }
+
+  void visitEnumElementDecl(EnumElementDecl *decl) {
+    auto params = decl->getParameterList();
+    if (!params) return;
+    for (auto param : params->getArray()) {
```

Comment:
ParameterList already has `begin`/`end`, so you should be able to drop the `getArray`.